File: pylib/anki/latex.py
# Copyright: Ankitects Pty Ltd and contributors
# License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html

from __future__ import annotations

import html
import os
import re
from dataclasses import dataclass
from typing import Any, List, Optional, Tuple

import anki
from anki import card_rendering_pb2, hooks
from anki.models import NotetypeDict
from anki.template import TemplateRenderContext, TemplateRenderOutput
from anki.utils import call, isMac, namedtmp, tmpdir

pngCommands = [
    ["latex", "-interaction=nonstopmode", "tmp.tex"],
    ["dvipng", "-D", "200", "-T", "tight", "tmp.dvi", "-o", "tmp.png"],
]

svgCommands = [
    ["latex", "-interaction=nonstopmode", "tmp.tex"],
    ["dvisvgm", "--no-fonts", "--exact", "-Z", "2", "tmp.dvi", "-o", "tmp.svg"],
]

build = True  # if off, use existing media but don't create new

# add standard tex install location to osx
if isMac:
    os.environ["PATH"] += ":/usr/texbin:/Library/TeX/texbin"


@dataclass
class ExtractedLatex:
    filename: str
    latex_body: str


@dataclass
class ExtractedLatexOutput:
    html: str
    latex: List[ExtractedLatex]

    @staticmethod
    def from_proto(
        proto: card_rendering_pb2.ExtractLatexResponse,
    ) -> ExtractedLatexOutput:
        return ExtractedLatexOutput(
            html=proto.text,
            latex=[
                ExtractedLatex(filename=l.filename, latex_body=l.latex_body)
                for l in proto.latex
            ],
        )


def on_card_did_render(
    output: TemplateRenderOutput, ctx: TemplateRenderContext
) -> None:
    output.question_text = render_latex(
        output.question_text, ctx.note_type(), ctx.col()
    )
    output.answer_text = render_latex(output.answer_text, ctx.note_type(), ctx.col())


def render_latex(
    html: str, model: NotetypeDict, col: anki.collection.Collection
) -> str:
    "Convert embedded latex tags in text to image links."
    html, err = render_latex_returning_errors(html, model, col)
    if err:
        html += "\n".join(err)
    return html


def render_latex_returning_errors(
    html: str,
    model: NotetypeDict,
    col: anki.collection.Collection,
    expand_clozes: bool = False,
) -> Tuple[str, List[str]]:
    """Returns (text, errors).

    errors will be non-empty if LaTeX failed to render."""
    svg = model.get("latexsvg", False)
    header = model["latexPre"]
    footer = model["latexPost"]

    proto = col._backend.extract_latex(text=html, svg=svg, expand_clozes=expand_clozes)
    out = ExtractedLatexOutput.from_proto(proto)
    errors = []
    html = out.html

    for latex in out.latex:
        # don't need to render?
        if not build or col.media.have(latex.filename):
            continue

        err = _save_latex_image(col, latex, header, footer, svg)
        if err is not None:
            errors.append(err)

    return html, errors


def _save_latex_image(
    col: anki.collection.Collection,
    extracted: ExtractedLatex,
    header: str,
    footer: str,
    svg: bool,
) -> Optional[str]:
    # add header/footer
    latex = f"{header}\n{extracted.latex_body}\n{footer}"
    # it's only really secure if run in a jail, but these are the most common
    tmplatex = latex.replace("\\includegraphics", "")
    for bad in (
        "\\write18",
        "\\readline",
        "\\input",
        "\\include",
        "\\catcode",
        "\\openout",
        "\\write",
        "\\loop",
        "\\def",
        "\\shipout",
    ):
        # don't mind if the sequence is only part of a command
        bad_re = f"\\{bad}[^a-zA-Z]"
        if re.search(bad_re, tmplatex):
            return col.tr.media_for_security_reasons_is_not(val=bad)

    # commands to use
    if svg:
        latexCmds = svgCommands
        ext = "svg"
    else:
        latexCmds = pngCommands
        ext = "png"

    # write into a temp file
    log = open(namedtmp("latex_log.txt"), "w")
    texpath = namedtmp("tmp.tex")
    texfile = open(texpath, "w", encoding="utf8")
    texfile.write(latex)
    texfile.close()
    oldcwd = os.getcwd()
    png_or_svg = namedtmp(f"tmp.{ext}")
    try:
        # generate png/svg
        os.chdir(tmpdir())
        for latexCmd in latexCmds:
            if call(latexCmd, stdout=log, stderr=log):
                return _errMsg(col, latexCmd[0], texpath)
        # add to media
        with open(png_or_svg, "rb") as file:
            data = file.read()
        col.media.write_data(extracted.filename, data)
        os.unlink(png_or_svg)
        return None
    finally:
        os.chdir(oldcwd)
        log.close()


def _errMsg(col: anki.collection.Collection, type: str, texpath: str) -> Any:
    msg = f"{col.tr.media_error_executing(val=type)}<br>"
    msg += f"{col.tr.media_generated_file(val=texpath)}<br>"
    try:
        with open(namedtmp("latex_log.txt", rm=False)) as f:
            log = f.read()
        if not log:
            raise Exception()
        msg += f"<small><pre>{html.escape(log)}</pre></small>"
    except:
        msg += col.tr.media_have_you_installed_latex_and_dvipngdvisvgm()
    return msg


def setup_hook() -> None:
    hooks.card_did_render.append(on_card_did_render)

File: Anki-Android/libanki/src/main/java/com/ichi2/anki/libanki/LaTeX.kt
/*
 * Copyright (c) 2009 Edu Zamora <edu.zasu@gmail.com>
 * Copyright (c) 2012 Kostas Spyropoulos <inigo.aldana@gmail.com>
 * Copyright (c) 2015 Houssam Salem <houssam.salem.au@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 3 of the License, or (at your option) any later
 * version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 * PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.ichi2.anki.libanki

import anki.card_rendering.ExtractLatexResponse
import anki.config.ConfigKey
import com.ichi2.anki.libanki.utils.LibAnkiAlias
import com.ichi2.anki.libanki.utils.append

data class ExtractedLatex(
    val fileName: String,
    val latexBody: String,
)

data class ExtractedLatexOutput(
    val html: String,
    val latex: List<ExtractedLatex>,
) {
    companion object {
        fun fromProto(proto: ExtractLatexResponse): ExtractedLatexOutput =
            ExtractedLatexOutput(
                html = proto.text,
                latex = proto.latexList.map { l -> ExtractedLatex(fileName = l.filename, latexBody = l.latexBody) },
            )
    }
}

@LibAnkiAlias("on_card_did_render")
fun onCardDidRender(
    col: Collection,
    output: TemplateManager.TemplateRenderContext.TemplateRenderOutput,
    ctx: TemplateManager.TemplateRenderContext,
) {
    output.questionText =
        renderLatex(
            output.questionText,
            ctx.noteType(),
            col,
        )
    output.answerText = renderLatex(output.answerText, ctx.noteType(), col)
}

/**
 * Convert embedded latex tags in text to image links.
 */
@LibAnkiAlias("render_latex")
fun renderLatex(
    html: String,
    model: NotetypeJson,
    col: Collection,
): String {
    val (html, err) = renderLatexRetuningErrors(html, model, col)
    val result = StringBuilder(html)
    if (err.isNotEmpty()) {
        result.append(err.joinToString("\n"))
    }
    return result.toString()
}

/**
 * Returns (text, errors).
 * errors will be non-empty if LaTeX failed to render.
 */
@LibAnkiAlias("render_latex_returning_errors")
fun renderLatexRetuningErrors(
    html: String,
    model: NotetypeJson,
    col: Collection,
    expandClozes: Boolean = false,
): Pair<String, List<String>> {
    val svg = model.latexsvg
    val header = model.latexPre
    val footer = model.latexPost

    val proto = col.backend.extractLatex(text = html, svg = svg, expandClozes = expandClozes)
    val out = ExtractedLatexOutput.fromProto(proto)
    val errors = mutableListOf<String>()
    val html = out.html
    val renderLatex = col.config.getBool(ConfigKey.Bool.RENDER_LATEX)

    for (latex in out.latex) {
        // don't need to render?
        if (col.media.have(latex.fileName)) {
            continue
        }
        if (!renderLatex) {
            errors.append(col.tr.preferencesLatexGenerationDisabled())
            return html to errors
        }

        val err = saveLatexImage(col, latex, header, footer, svg)
        if (err != null) {
            errors.append(err)
        }
    }
    return html to errors
}

// deliberately set the method to always return null as
// AnkiDroid does not support the generation of LaTeX images.
@Suppress("unused")
@LibAnkiAlias("_save_latex_image")
private fun saveLatexImage(
    col: Collection,
    extractedLatex: ExtractedLatex,
    header: String,
    footer: String,
    svg: Boolean,
): String? = null
